README.md
# battery-entity-row

Show battery states or attributes with dynamic icon on entity rows in Home Assistant's Lovelace UI

[![GH-release](https://img.shields.io/github/v/release/benct/lovelace-battery-entity-row.svg?style=flat-square)](https://github.com/benct/lovelace-battery-entity-row/releases)
[![GH-downloads](https://img.shields.io/github/downloads/benct/lovelace-battery-entity-row/total?style=flat-square)](https://github.com/benct/lovelace-battery-entity-row/releases)
[![GH-last-commit](https://img.shields.io/github/last-commit/benct/lovelace-battery-entity-row.svg?style=flat-square)](https://github.com/benct/lovelace-battery-entity-row/commits/master)
[![GH-code-size](https://img.shields.io/github/languages/code-size/benct/lovelace-battery-entity-row.svg?color=red&style=flat-square)](https://github.com/benct/lovelace-battery-entity-row)
[![hacs_badge](https://img.shields.io/badge/HACS-Default-orange.svg?style=flat-square)](https://github.com/hacs)

Rewritten and improved version of cbulock's [battery-entity](https://github.com/cbulock/lovelace-battery-entity) card _(deprecated/unmaintained)_.

**NOTE:** This is not a standalone lovelace card, but a row element for the [entities](https://www.home-assistant.io/lovelace/entities/) card.
If you need a standalone card or want _a lot_ more customizability, check out maxwroc's [battery-state-card](https://github.com/maxwroc/battery-state-card).

## Installation

Manually add [battery-entity-row.js](https://raw.githubusercontent.com/benct/lovelace-battery-entity-row/master/battery-entity-row.js)
to your `<config>/www/` folder and add the following to the `configuration.yaml` file:

```yaml
lovelace:
  resources:
    - url: /local/battery-entity-row.js?v=1.3.1
      type: module
```

_OR_ install using [HACS](https://hacs.xyz/) and add this (if in YAML mode):

```yaml
lovelace:
  resources:
    - url: /hacsfiles/lovelace-battery-entity-row/battery-entity-row.js
      type: module
```

The above configuration can be managed directly in the Configuration -> Lovelace Dashboards -> Resources panel when not using YAML mode,
or added by clicking the "Add to lovelace" button on the HACS dashboard after installing the plugin.

## Configuration

This card produces an `entity-row` and must therefore be configured as an entity in an [entities](https://www.home-assistant.io/lovelace/entities/) card.

The battery level value is fetched from the entity `state`, from the attribute `battery` or `battery_level`,
or from a custom attribute defined with the `attribute` option. Numeric values (`0-100`) and some predefined
string values (`high`, `normal`, `low`, etc..) are supported as a battery level value.

| Name            | Type        | Default         | Description                                                                    |
| --------------- | ----------- | --------------- | ------------------------------------------------------------------------------ |
| type            | string      | **Required**    | `custom:battery-entity-row`                                                    |
| entity          | string      | **Required**    | `domain.my_entity_id`                                                          |
| attribute       | string      | `battery_level` | Override battery level attribute                                               |
| name            | string      | `friendly_name` | Override entity `friendly_name`                                                |
| secondary\_info | string      |                 | `last-changed`, `last-updated` or an attribute of the entity.                  |
| unit            | string/bool | `%`             | Override default `unit`, or hide with `false`                                  |
| icon            | string      |                 | Override dynamic battery `icon`                                                |
| warning         | number      | `35`            | Level at which the icon will appear yellow                                     |
| critical        | number      | `15`            | Level at which the icon will appear red                                        |
| charging        | bool/object | `false`         | Indicate charging based on entity state. See charging object for more options. |

Currently limited support for `secondary_info` option with value `last-changed`.

### Charging object

| Name      | Type        | Default              | Description                                          |
| --------- | ----------- | -------------------- | ---------------------------------------------------- |
| entity    | string      | `main entity`        | Get charging state from another entity               |
| attribute | string      |                      | Get charging state from an attribute                 |
| state     | string/list | `"on"`, `"charging"` | Add values that indicate charging (case insensitive) |

## Examples

![battery-entity-row](https://raw.githubusercontent.com/benct/lovelace-battery-entity-row/master/example.png)

```yaml
type: entities
entities:
  - type: custom:battery-entity-row
    entity: sensor.bedroom_temperature

  - type: custom:battery-entity-row
    entity: sensor.bedroom_temperature
    attribute: battery_percent
    name: Some battery
    unit: percent
    icon: mdi:battery-alert
    secondary_info: last-changed
    warning: 50
    critical: 25
    charging: true

  - type: custom:battery-entity-row
    entity: sensor.bedroom_temperature
    name: Charging battery
    charging:
      entity: binary_sensor.bedroom_temperature_charger
      attribute: charging
      state:
        - Enabled
        - is_charging
```

Usage in [auto-entities](https://github.com/thomasloven/lovelace-auto-entities) card:

```yaml
type: custom:auto-entities
card:
  type: entities
filter:
  include:
    - entity_id: sensor.battery*   # or use other matchers
      options:
        type: custom:battery-entity-row
        <battery-entity-row options>
```

## My cards

[xiaomi-vacuum-card](https://github.com/benct/lovelace-xiaomi-vacuum-card) |
[multiple-entity-row](https://github.com/benct/lovelace-multiple-entity-row) |
[github-entity-row](https://github.com/benct/lovelace-github-entity-row) |
[battery-entity-row](https://github.com/benct/lovelace-battery-entity-row) |
[~~attribute-entity-row~~](https://github.com/benct/lovelace-attribute-entity-row)

[![BMC](https://www.buymeacoffee.com/assets/img/custom_images/white_img.png)](https://www.buymeacoff.ee/benct)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## 1.3.1

**Fixed:**
- Incorrect over-simplification of secondary_info handling

## 1.3.0

**Added:**
- Support `last-updated` on secondary_info (#10)
- Support custom entity attributes on secondary_info (#10)

**Fixed:**
- Breaking change to relative datetime objects (#9)

## 1.2.0

**Added:**
- Support some predefined string values (#5)
- Support custom charging state or attribute values (#7)

**Changed:**
- Prevent unnecessary re-rendering with `shouldUpdate` function (#8)
- Refactor and improve configuration handling

## 1.1.0

**Added:**
- Support for battery charging state or attribute (#4)

**Changed:**
- Improve unknown/unavailable battery state display
- Allow disabling/hiding `unit` with value `false`

## 1.0.0

- **Initial release**


battery-entity-row.js
((LitElement) => {
    console.info(
        '%c BATTERY-ENTITY-ROW %c 1.3.0 ',
        'color: cyan; background: black; font-weight: bold;',
        'color: darkblue; background: white; font-weight: bold;',
    );
    const {html, css} = LitElement.prototype;

    const defaultOnStates = ['on', 'charging', 'true'];

    class BatteryEntityRow extends LitElement {

        static get properties() {
            return {
                _hass: Object,
                _config: Object,
                stateObj: Object
            }
        }

        static get styles() {
            return css`
          :host {
            display: flex;
            align-items: center;
          }
          .flex {
            flex: 1;
            margin-left: 16px;
            display: flex;
            justify-content: space-between;
            align-items: center;
            min-width: 0;
          }
          .info, state-badge {
            cursor: pointer;
          }
          .secondary {
            color: var(--secondary-text-color);
          }
          .good {
            color: var(--label-badge-green);
          }
          .warning {
            color: var(--label-badge-yellow);
          }
          .critical {
            color: var(--label-badge-red);
          }`;
        }

        render() {
            if (!this._hass || !this._config) return html``;
            if (!this.stateObj) return this.renderWarning();

            const charging = this.getChargingState(this._config.charging)
            const batteryValue = this.getBatteryLevel(this._config.attribute);

            const isUnavailable = !batteryValue || ['unavailable', 'unknown'].includes(batteryValue);
            const isNumeric = !isNaN(parseFloat(batteryValue)) && isFinite(batteryValue);

            const numericValue = isUnavailable ? null : isNumeric ? batteryValue : this.parseStringValue(batteryValue);

            const icon = this._config.icon || this.getIcon(numericValue, charging);
            const color = this.getColor(numericValue);

            const name = this._config.name || this.stateObj.attributes.friendly_name;
            const unit = this._config.unit === false ? null : (this._config.unit || (isNumeric ? '%' : null));
            const state = isUnavailable
                ? this._hass.localize('state.default.unknown')
                : html`${batteryValue}${unit && html`&nbsp;${unit}`}`;

            return html`
            <state-badge
                .stateObj="${this.stateObj}"
                .overrideIcon="${icon}"
                @click="${this.moreInfo}"
                class="pointer ${color}">
            </state-badge>
            <div class="flex" @click="${this.moreInfo}">
                <div class="info">${name}${this.renderSecondaryInfo()}</div>
                <div class="state">${state}</div>
            </div>`;
        }

        renderSecondaryInfo() {
            const secondaryInfo = this._config.secondary_info;
            let content = undefined;

            if (secondaryInfo === 'last-changed') {
                content = html`<ha-relative-time .datetime="${this.stateObj.last_changed}" .hass="${this._hass}"></ha-relative-time>`;
            } else if (secondaryInfo === 'last-updated') {
                content = html`<ha-relative-time .datetime="${this.stateObj.last_updated}" .hass="${this._hass}"></ha-relative-time>`;
            } else if (secondaryInfo in this.stateObj.attributes) {
                content = this.stateObj.attributes[secondaryInfo];
            }
            return content ? html`<div class="secondary">${content}</div>` : null;
        }

        renderWarning() {
            return html`<hui-warning>
                ${this._hass.localize('ui.panel.lovelace.warning.entity_not_found', 'entity', this._config.entity)}
            </hui-warning>`;
        }

        setConfig(config) {
            if (!config.entity) throw new Error('Please define a valid entity.');

            this.moreInfo = () => this.fireEvent(this, 'hass-more-info', {entityId: config.entity});
            this._config = config;
        }

        shouldUpdate(changedProps) {
            return changedProps.has('stateObj');
        }

        set hass(hass) {
            this._hass = hass;

            if (hass && this._config) {
                this.stateObj = this._config.entity in hass.states ? hass.states[this._config.entity] : null;
            }
        }

        getBatteryLevel(attribute) {
            let batteryValue = this.stateObj.state;
            if (this.stateObj.attributes.battery) batteryValue = this.stateObj.attributes.battery;
            if (this.stateObj.attributes.battery_level) batteryValue = this.stateObj.attributes.battery_level;
            if (this.stateObj.attributes[attribute]) batteryValue = this.stateObj.attributes[attribute];
            return !isNaN(parseFloat(batteryValue)) && isFinite(batteryValue)
                ? Math.round(parseInt(batteryValue, 10)) : batteryValue;
        }

        getChargingState(chargingConfig) {
            if (!chargingConfig) return false;
            if (chargingConfig === true) {
                return defaultOnStates.includes(this.stateObj.state.toString().toLowerCase());
            }

            const additionalStates = chargingConfig.state || [];
            const onStates = defaultOnStates.concat(additionalStates).map(value => value.toString().toLowerCase());

            const entity = (chargingConfig.entity && chargingConfig.entity in this._hass.states)
                ? this._hass.states[chargingConfig.entity] : this.stateObj;
            const state = chargingConfig.attribute ? entity.attributes[chargingConfig.attribute] : entity.state;
            return onStates.includes(state.toString().toLowerCase());
        }

        getIcon(batteryLevel, charging) {
            if (!batteryLevel) return 'mdi:battery-unknown';
            const roundedLevel = Math.round(batteryLevel / 10) * 10;
            return roundedLevel >= 100
                ? (charging ? 'mdi:battery-charging-100' : 'mdi:battery')
                : roundedLevel === 0
                    ? (charging ? 'mdi:battery-charging-outline' : 'mdi:battery-outline')
                    : (charging ? 'mdi:battery-charging-' : 'mdi:battery-') + roundedLevel;
        }

        getColor(batteryLevel) {
            if (!batteryLevel) return 'unknown';
            const warning = this._config.warning || 35;
            const critical = this._config.critical || 15;
            return (batteryLevel > warning)
                ? 'good'
                : (batteryLevel > critical)
                    ? 'warning'
                    : 'critical';
        }

        parseStringValue(v) {
            const val = v.toString().toLowerCase();
            if (['full', 'high', 'max', 'maximum'].includes(val)) return 90;
            if (['medium', 'med', 'normal'].includes(val)) return 50;
            if (['low', 'min', 'minimal'].includes(val)) return 10;
            if (['empty', 'critical', 'none'].includes(val)) return 0;
            return null;
        }

        fireEvent(node, type, detail = {}, options = {}) {
            const event = new Event(type, {
                bubbles: options.bubbles || true,
                cancelable: options.cancelable || true,
                composed: options.composed || true,
            });
            event.detail = detail;
            node.dispatchEvent(event);
        }
    }

    customElements.define('battery-entity-row', BatteryEntityRow);
})(window.LitElement || Object.getPrototypeOf(customElements.get('hui-masonry-view') || customElements.get('hui-view')));
